### ISSUE TITLE
Follow Mode does not handle file creation correctly

### ISSUE THREAD

**Issue by tobous** [saros-project/saros#445]
When the followee creates a new file during the follow mode, the follower does not open the file automatically.

The follower is throwing an exception internally.
<details><summary>Exception</summary>
<p>

```Java
ERROR 16:29:37,857 [AWT-EventQueue-0] (SarosSession.java:195) error while invoking active activity consumer: de.fu_berlin.inf.dpp.editor.FollowModeManager$3@7c635aa0, activity: EditorActivity(type: ACTIVATED, path: SPath [project=IntelliJProjectImpl : SharedModule, projectRelativePath=src/plot_ha/other/test/ha/Test3.java])
java.lang.IllegalArgumentException: Argument for @NotNull parameter 'resource' of de/fu_berlin/inf/dpp/intellij/session/SessionUtils.isShared must not be null
	at de.fu_berlin.inf.dpp.intellij.session.SessionUtils.$$$reportNull$$$0(SessionUtils.java)
	at de.fu_berlin.inf.dpp.intellij.session.SessionUtils.isShared(SessionUtils.java)
	at de.fu_berlin.inf.dpp.intellij.editor.LocalEditorManipulator.openEditor(LocalEditorManipulator.java:67)
	at de.fu_berlin.inf.dpp.intellij.editor.EditorManager.lambda$openEditor$3(EditorManager.java:836)
	at com.intellij.openapi.application.impl.ApplicationImpl.invokeAndWait(ApplicationImpl.java:652)
	at de.fu_berlin.inf.dpp.intellij.editor.EditorManager.executeInUIThreadSynchronous(EditorManager.java:812)
	at de.fu_berlin.inf.dpp.intellij.editor.EditorManager.openEditor(EditorManager.java:836)
	at de.fu_berlin.inf.dpp.editor.FollowModeManager$3.receive(FollowModeManager.java:118)
	at de.fu_berlin.inf.dpp.activities.EditorActivity.dispatch(EditorActivity.java:106)
	at de.fu_berlin.inf.dpp.session.AbstractActivityConsumer.exec(AbstractActivityConsumer.java:81)
	at de.fu_berlin.inf.dpp.editor.FollowModeManager$3.exec(FollowModeManager.java:106)
	at de.fu_berlin.inf.dpp.session.internal.SarosSession$2.execute(SarosSession.java:193)
	at de.fu_berlin.inf.dpp.session.internal.ActivityHandler$3.run(ActivityHandler.java:339)
	at de.fu_berlin.inf.dpp.util.ThreadUtils$1.run(ThreadUtils.java:35)
	at com.intellij.openapi.application.TransactionGuardImpl$2.run(TransactionGuardImpl.java:315)
	at com.intellij.openapi.application.impl.LaterInvocator$1.run(LaterInvocator.java:156)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.doRun(LaterInvocator.java:447)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.runNextEvent(LaterInvocator.java:431)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.run(LaterInvocator.java:415)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:781)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:722)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:382)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```

</p>
</details>

My guess would be that the follower is trying to access/open the file before it was created locally. This still has to be investigated.

**Comment by tobous** [comment]
The issue seems to be that the activities are dispatched in the wrong order on the followee side: The activity for the new active editor (of the created file) is dispatched before the activity for the file creation.

The events are triggered in the correct order by IntelliJ (first the listener for file creations is triggered and then the listener for editor selection changes is triggered). The problem is that the different listener notifications are *executed* in parallel. The [logic to process the file creation](https://github.com/saros-project/saros/blob/master/de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/project/LocalFilesystemModificationHandler.java#L211) seems to take longer than the [logic to process the editor status change](https://github.com/saros-project/saros/blob/master/de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/editor/LocalEditorStatusChangeHandler.java#L205), which causes the corresponding Saros activities to be dispatched in the wrong order.

**Comment by tobous** [comment]
I am not sure how to solve this properly.

One option would be to handle this with a queuing logic on the receiving side, just like [viewport adjustments](https://github.com/saros-project/saros/commit/1a6a283161eb08ef6b494cff334e6ad1c5b8d966) (where the activities are triggered in the "wrong" order by IntelliJ). But since the logic is [located in the core](https://github.com/saros-project/saros/blob/master/de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/editor/FollowModeManager.java#L118), there is no good place to add such a queuing logic that doesn't influence other Saros implementations unnecessarily.

Another option would be to enforce the order that the listeners were triggered in by IntelliJ through some sort of synchronization mechanism.
But the only solution I can think of is to create runnables in the different handlers (instead of executing the needed logic directly) and then pass these runnables to an executor that executes them sequentially on a separate thread (so that the handlers aren't blocked).
The downsides to this approach are that it would reduce performance/throughput, make the logic harder to understand, and add  an extra level of indentation to most of the handler logic, thereby making it harder to read.

The easiest but also most hacky solution I can think of it to just add a nullcheck to the open editor method that drops the activity and to just create an editor activity "manually" as of the file creation handler to enforce the correct order. But this would only solve this specific case and I don't know whether there are more cases similar to this one.

**Comment by srossbach** [comment]
I guess you are running into this problem https://www.jetbrains.org/intellij/sdk/docs/basics/virtual_file_system.html#synchronous-and-asynchronous-refreshes

i.e the file is created outside the VFS and is then synchronized via a refresh ?

**Comment by tobous** [comment]
No, the problem is that all listeners are executed in parallel. When a file is created, the file creation listener is triggered first. Then, as the created file is also opened by default, the editor selection listener is triggered. The problem is that the file creation handler runs longer than the editor selection handler, meaning the editor activated activity is dispatched before the file creation activity.

When the other participant is in the follow mode, this leads to a situation where saros tries to open an  editor for a file that does not exist locally.

**Comment by srossbach** [comment]
You could also create an internal buffer and wait for the file arrival, i.e the affected component also needs to listen to file events. I do not know if this is possible as we only have two levels. Maybe you want to have a look at this too https://saros-build.imp.fu-berlin.de/gerrit/#/c/3117/

**Comment by srossbach** [comment]
I rethought my solution and I would not suggest it as this can cause issues in the core because the core is coded that way that when an activity arrives or is generated it assumes that there was no parallel race conditions (especially on the file system level).

**Comment by srossbach** [comment]
As stated here: https://www.jetbrains.org/intellij/sdk/docs/basics/virtual_file_system.html

All changes happening in the virtual file system, either as a result of refresh operations or caused by user’s actions, are reported as virtual file system events. VFS events are always fired in the event dispatch thread, and in a write action.

Sourcecode:

@Override
  public void notifyPropertyChanged(@NotNull VirtualFile virtualFile, @VirtualFile.PropName @NotNull String property, Object oldValue, Object newValue) {
    Application app = ApplicationManager.getApplication();
    app.invokeLater(() -> {
      if (virtualFile.isValid() && !app.isDisposed()) {
        app.runWriteAction(() -> {
          List<VFileEvent> events = Collections.singletonList(new VFilePropertyChangeEvent(this, virtualFile, property, oldValue, newValue, false));
          BulkFileListener listener = app.getMessageBus().syncPublisher(VirtualFileManager.VFS_CHANGES);
          listener.before(events);
          listener.after(events);
        });
      }
    }, ModalityState.NON_MODAL);

I have looked at the source files and this is actually the case. Now I am wondering which thread is also running "in parallel". Could it not be the case that when you create a file (as you see above the anouncement is delayed) in the ApplicationThread the announcement comes after the creation and so the editor listener is triggered always first.

I mean you could try the following, when you receive an editor open (first check if the file is present but I guess that is the case.) Then you call invokeLater and in that Runnable you call invokeLater again (so you are now queued behind the VFS broadcast, maybe one invokeLater is sufficent) and then fire the event.

On the other hand if you open an editor for a file that already exists you are delaying too much.

IIRC there was an hack that told the "ResourceManager" to ignore certain file changes if a file was created from a template, because the EditorManager was sending this by itself. Did you removed this hack ?

**Comment by tobous** [comment]
### Update:

Wow... I found the issue. It was a problem of my own making. The problem is in the file creation handling [line 263](https://github.com/saros-project/saros/blob/pr/intellij/fix_context_separation/intellij/src/saros/intellij/eventhandler/filesystem/LocalFilesystemModificationHandler.java#L263):

I call `openEditor(...)` for the newly created file to register it with the editor pool, to ensure that the initial file activities (like the initial content addition) are recognized by Saros as valid activities.

But, by calling the method manually before the file creation activity is dispatched, I trigger the editor selection change listener prematurely, thereby causing the issue.

I have two options to fix this:

1. By moving the call to `openEditors` (including a check whether the resource is actually a file) behind the creation activity dispatching, the issue is avoided.

2. By removing the call to `openEditor`, the issue is avoided and the rest of the logic still seems to works as it should (including the initial content transmission). The editor is still registered to the editor pool indirectly/"by accident" through the [`PreexistingSelectionDispatcher`](https://github.com/saros-project/saros/blob/pr/intellij/fix_context_separation/intellij/src/saros/intellij/eventhandler/editor/editorstate/PreexistingSelectionDispatcher.java#L64) and [`AnnotationUpdater`](https://github.com/saros-project/saros/blob/pr/intellij/fix_context_separation/intellij/src/saros/intellij/eventhandler/editor/editorstate/AnnotationUpdater.java#L71). But this should probably be done explicitly somewhere to ensure that is always happens, even when the other handlers are removed/modified. Will either add it to [`EditorStatusChangeActivityDispatcher`](https://github.com/saros-project/saros/blob/pr/intellij/fix_context_separation/intellij/src/saros/intellij/eventhandler/editor/editorstate/EditorStatusChangeActivityDispatcher.java) or create a new handler for it.

I still have to check whether opening the editor inside the file creation handler at all causes problems, but otherwise, the first solution would probably be preferable.

This somewhat annoys me as I have invested a lot of time into this trying multiple synchronization methods. But whatever.
I will still have a look at how editor activities are processed as this is probably useful knowledge for future work. Will also check all handlers for similar issues.

-----
### Old message:

> Could it not be the case that when you create a file (as you see above the anouncement is delayed) in the ApplicationThread the announcement comes after the creation and so the editor listener is triggered always first.

No the activities are triggered in the correct order. I tested it again just now by adding a logging instruction to the beginning and end of the [resource creation handling](https://github.com/saros-project/saros/blob/pr/intellij/fix_context_separation/intellij/src/saros/intellij/eventhandler/filesystem/LocalFilesystemModificationHandler.java) ([after l. 231](https://github.com/saros-project/saros/blob/pr/intellij/fix_context_separation/intellij/src/saros/intellij/eventhandler/filesystem/LocalFilesystemModificationHandler.java#L231) and [after l. 266](https://github.com/saros-project/saros/blob/pr/intellij/fix_context_separation/intellij/src/saros/intellij/eventhandler/filesystem/LocalFilesystemModificationHandler.java#L266)) and [editor selection change handling](https://github.com/saros-project/saros/blob/pr/intellij/fix_context_separation/intellij/src/saros/intellij/eventhandler/editor/editorstate/EditorStatusChangeActivityDispatcher.java) (before and after [l.64](https://github.com/saros-project/saros/blob/pr/intellij/fix_context_separation/intellij/src/saros/intellij/eventhandler/editor/editorstate/EditorStatusChangeActivityDispatcher.java#L64)). This consistently results in the following order:
```
FATAL 17:59:35,381 [AWT-EventQueue-0] (LocalFilesystemModificationHandler.java:234) Reacting to file creation for file file:///home/tobous/IdeaProjects/HostTestProject/SharedModule/src/new_base/TestClass.java
[...]
FATAL 17:59:35,434 [AWT-EventQueue-0] (EditorStatusChangeActivityDispatcher.java:66) Reacting to editor activation for file:///home/tobous/IdeaProjects/HostTestProject/SharedModule/src/new_base/TestClass.java
FATAL 17:59:35,435 [AWT-EventQueue-0] (EditorStatusChangeActivityDispatcher.java:68) Dispatched editor activation activity for file:///home/tobous/IdeaProjects/HostTestProject/SharedModule/src/new_base/TestClass.java
[...]
FATAL 17:59:35,449 [AWT-EventQueue-0] (LocalFilesystemModificationHandler.java:271) Dispatched file creation activity for file file:///home/tobous/IdeaProjects/HostTestProject/SharedModule/src/new_base/TestClass.java
```

> IIRC there was an hack that told the "ResourceManager" to ignore certain file changes if a file was created from a template, because the EditorManager was sending this by itself. Did you removed this hack ?

I think you mean #80. Yes, I removed the hack with the introduction of the reworked handling of local filesystem activities. But I couldn't recreate the issue. The files are initially empty and the content is added later on after the creation. This addition of the template content is picked up by the document modification listener and is therefore correctly send to the other session participants.
